pkg/crawler/central/central.go

```diff
@@ -0,0 +1,376 @@
+// Therefore, this implementation follows a two-step process:
+//
+//   1. Extract a list of GAVC (GroupID, ArtifactID, Version, Classifier) coordinates from the central index.
+//   2. For each GAVC, fetch the correct SHA1 value directly from Maven Central repository.
+//
+// This ensures that only accurate SHA1s are stored and used.
+//
+// The overall flow can be visualized as:
+//
+//   +-------------------+         +-------------------+
+//   | Central Index     |         | Maven Central     |
+//   |  (GAVC list)     |         |  (SHA1 endpoint)  |
+//   +--------+----------+         +---------+---------+
+//            |                              ^
+//            | 1. Extract GAVC list         |
+//            v                              |
+//   +-------------------+                   |
+//   |   This Crawler    |-------------------+
+//   |                   |   2. Fetch SHA1 for each GAVC
+//   +-------------------+
+//
+// See: https://github.com/aquasecurity/trivy-java-db/pull/58#issuecomment-2890441162
+
+// Source implements the crawler.Source interface for Maven Central Index
+type Source struct {
+	indexPath   string
+	url         string
+	httpClient  *retryablehttp.Client
+	logger      *slog.Logger
+	storedGAVCs map[uint64]struct{} // GroupID, ArtifactID, Version, Classifier
+	records     []types.Record      // All GAVC records stored in central index
+	processed   int
+	errCount    int
+}
+
+// New creates a new Maven Central Index source
+func New(httpClient *retryablehttp.Client, url string, cacheDir string, storedGAVs map[uint64]struct{}) (*Source, error) {
+	if cacheDir == "" {
+		return nil, xerrors.New("cache directory not specified")
+	}
+
+	targetURL := cmp.Or(url, defaultURL)
+	fileName := path.Base(targetURL)
+	// Ensure the cache directory exists
+	centralCacheDir := filepath.Join(cacheDir, "central-index")
+	if err := os.MkdirAll(centralCacheDir, 0755); err != nil {
+		return nil, xerrors.Errorf("failed to create cache directory: %w", err)
+	}
+
+	return &Source{
+		indexPath:   filepath.Join(centralCacheDir, fileName),
+		url:         targetURL,
+		httpClient:  httpClient,
+		logger:      slog.Default().With(slog.String("source", "central")),
+		storedGAVCs: storedGAVs,
+	}, nil
+}
+
+// downloadIndex downloads the index file to a cached location and returns the path
+func (s *Source) downloadIndex() (err error) {
+	// Check if the file already exists in cache
+	if _, err := os.Stat(s.indexPath); err == nil {
+		s.logger.Info("Using cached Maven Central Index", slog.String("path", s.indexPath))
+		return nil
+	}
+
+	s.logger.Info("Downloading Maven Central Index", slog.String("url", s.url), slog.String("path", s.indexPath))
+
+	// Create the output file
+	outFile, err := os.Create(s.indexPath)
+	if err != nil {
+		return xerrors.Errorf("failed to create output file: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(s.indexPath)
+		}
+	}()
```

Comment:
IIUC we don't need defer here.
```suggestion
	if err != nil {
		os.Remove(s.indexPath)
		return xerrors.Errorf("failed to create output file: %w", err)
	}
```

Reply:
It is also necessary to delete it in case of failures in subsequent processes, such as in `io.Copy`.

Reply:
Got it!
Thanks for the explanation.